test/PushPullTests.cpp

```diff
@@ -0,0 +1,460 @@
+#include "PushPullTests.h"
+#include "Death.h"
+#include "FileIO.h"
+#include <g3log/g3log.hpp>
+#include <functional>
+namespace {
+   std::string GetTcpLocation() {
+      int max = 9000;
+      int min = 7000;
+      int port = (rand() % (max - min)) + min;
+      std::string tcpLocation("tcp://127.0.0.1:");
+      tcpLocation.append(std::to_string(port));
+      return tcpLocation;
+   }
+
+   std::string GetIpcLocation() {
+      int pid = getpid();
+      std::string ipcLocation("ipc:///tmp/");
+      ipcLocation.append("PushPullTests");
+      ipcLocation.append(std::to_string(pid));
+      ipcLocation.append(".ipc");
+      return ipcLocation;
+   }
+
+   std::vector<std::shared_ptr<Push>> CreatePushs(int howManyPushs,
+                                                  int howManyPulls,
+                                                  const std::string& location) {
+      std::vector<std::shared_ptr<Push>> toReturn;
+      for (int i = 0; i < howManyPushs ; i ++) {
+         if (howManyPushs == 1 && howManyPulls == 1) {
+            // one to one, use default behavior (connect)
+            LOG(DEBUG) << "one to one push setup";
+            toReturn.push_back(std::make_shared<Push>(location));
+         } else if (howManyPushs > 1) {
```

Comment:
I've problem reading this code and understanding the outcome. Same with the CreatePulls below. 

I think it would be much clearer if you had separate ```Create``` functions.  A hint that you also thought it was complex are the comments for each option. 

Why not create declarative function names instead? (this function ---> 4 functions)